Trying to create Windows nodegroup with 1.15 causes SIGSEGV

**What happened?**
A description of actual behavior (with error messages).

I am trying to create a Windows nodegroup. I followed the instructions on the official documentation: https://docs.aws.amazon.com/eks/latest/userguide/launch-windows-workers.html

I ran this command:

$ eksctl create nodegroup --cluster training --profile mfa --region us-east-1 --name windows -m 1 -M 3 -N 2 --node-ami-family WindowsServer2019FullContainer --node-type m5.large

And got this error:

[ℹ]  eksctl version 0.15.0
[ℹ]  using region us-east-1
[ℹ]  1 nodegroup(s) that already exist (ng-b392aa65) will be excluded
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1844eb0]

goroutine 1 [running]:
github.com/weaveworks/eksctl/pkg/ami.Use(0x33eabc0, 0xc00000ea88, 0xc0003242c0, 0x7ffc460ed7bc, 0x4)
	github.com/weaveworks/eksctl@/pkg/ami/api.go:57 +0x160
github.com/weaveworks/eksctl/pkg/eks.EnsureAMI(0x33ad7e0, 0xc0004d7220, 0x7ffc460ed7bc, 0x4, 0xc0003242c0, 0x0, 0x0)
	github.com/weaveworks/eksctl@/pkg/eks/api.go:301 +0x361
github.com/weaveworks/eksctl/pkg/ctl/create.doCreateNodeGroups(0xc00029e280, 0xc0003242c0, 0x1, 0x0, 0x0)
	github.com/weaveworks/eksctl@/pkg/ctl/create/nodegroup.go:127 +0x3db
github.com/weaveworks/eksctl/pkg/ctl/create.createNodeGroupCmd.func1(0xc00030ca00, 0xc00025a3c0, 0x0, 0x14, 0x0, 0x0)
	github.com/weaveworks/eksctl@/pkg/ctl/create/nodegroup.go:40 +0xa5
github.com/spf13/cobra.(*Command).execute(0xc00030ca00, 0xc00025a280, 0x14, 0x14, 0xc00030ca00, 0xc00025a280)
	github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc00030c280, 0xc0002b1c20, 0x1, 0x1)
	github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
	github.com/weaveworks/eksctl@/cmd/eksctl/main.go:79 +0x407

When I added --version 1.14, everything created successfully, but using 1.15 does not work at all.

To reproduce, all I did was run the "install-vpc-controllers" command from the documentation above and then the command. The cluster itself was default.

**Anything else we need to know?**
What OS are you using, are you using a downloaded binary or did you compile eksctl, what type of AWS credentials are you using (i.e. default/named profile, MFA) - please don't include actual credentials though!

I'm on Ubuntu 18.04.4 with eksctl 0.15.0

**Versions**
Please paste in the output of these commands:
```
$ eksctl version
$ kubectl version
```lrelease

**Logs**
Include the output of the command line when running eksctl. If possible, eksctl should be run with debug logs. For example:
`eksctl get clusters -v 4`
Make sure you redact any sensitive information before posting.
If the output is long, please consider a Gist.

--- a/pkg/ami/ami_test.go
+++ b/pkg/ami/ami_test.go
@@ -1,13 +1,14 @@
-package ami
+package ami_test
 
 import (
-	"strconv"
+	"fmt"
 	"strings"
 	"testing"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/stretchr/testify/mock"
+	"github.com/weaveworks/eksctl/pkg/ami"
 	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
 	"github.com/weaveworks/eksctl/pkg/testutils/mockprovider"
 )
@@ -16,8 +17,10 @@ func TestUseAMI(t *testing.T) {
 	amiTests := []struct {
 		blockDeviceMappings []*ec2.BlockDeviceMapping
 		rootDeviceName      string
+		description         string
 	}{
 		{
+			description:    "Root device mapping not at index 0 (Windows AMIs in some regions)",
 			rootDeviceName: "/dev/sda1",
 			blockDeviceMappings: []*ec2.BlockDeviceMapping{
 				{
@@ -36,12 +39,44 @@ func TestUseAMI(t *testing.T) {
 				},
 			},
 		},
+		{
+			description:    "Only one device mapping (AL2 AMIs)",
+			rootDeviceName: "/dev/sda1",
+			blockDeviceMappings: []*ec2.BlockDeviceMapping{
+				{
+					DeviceName: aws.String("/dev/sda1"),
+					Ebs: &ec2.EbsBlockDevice{
+						Encrypted: aws.Bool(true),
+					},
+				},
+			},
+		},
+		{
+			description:    "Different root device name",
+			rootDeviceName: "/dev/xvda",
+			blockDeviceMappings: []*ec2.BlockDeviceMapping{
+				{
+					DeviceName:  aws.String("xvdca"),
+					VirtualName: aws.String("ephemeral0"),
+				},
+				{
+					DeviceName: aws.String("/dev/xvda"),
+					Ebs: &ec2.EbsBlockDevice{
+						Encrypted: aws.Bool(true),
+					},
+				},
+				{
+					DeviceName:  aws.String("xvdcb"),
+					VirtualName: aws.String("ephemeral1"),
+				},
+			},
+		},
 	}
 
 	for i, tt := range amiTests {
-		t.Run(strconv.Itoa(i), func(t *testing.T) {
+		t.Run(fmt.Sprintf("%d: %s", i, tt.description), func(t *testing.T) {
 			mockProvider := mockDescribeImages(tt.blockDeviceMappings, tt.rootDeviceName)
-			err := Use(mockProvider.MockEC2(), &api.NodeGroup{
+			err := ami.Use(mockProvider.MockEC2(), &api.NodeGroup{
 				AMI: "ami-0121d8347f8191f90",
 			})
 

--- a/pkg/ami/ami_test.go
+++ b/pkg/ami/ami_test.go
@@ -18,6 +18,8 @@ func TestUseAMI(t *testing.T) {
 		blockDeviceMappings []*ec2.BlockDeviceMapping
 		rootDeviceName      string
 		description         string
+
+		encrypted bool
 	}{
 		{
 			description:    "Root device mapping not at index 0 (Windows AMIs in some regions)",
@@ -38,6 +40,8 @@ func TestUseAMI(t *testing.T) {
 					},
 				},
 			},
+
+			encrypted: true,
 		},
 		{
 			description:    "Only one device mapping (AL2 AMIs)",
@@ -50,6 +54,8 @@ func TestUseAMI(t *testing.T) {
 					},
 				},
 			},
+
+			encrypted: true,
 		},
 		{
 			description:    "Different root device name",
@@ -62,7 +68,7 @@ func TestUseAMI(t *testing.T) {
 				{
 					DeviceName: aws.String("/dev/xvda"),
 					Ebs: &ec2.EbsBlockDevice{
-						Encrypted: aws.Bool(true),
+						Encrypted: aws.Bool(false),
 					},
 				},
 				{
@@ -70,18 +76,25 @@ func TestUseAMI(t *testing.T) {
 					VirtualName: aws.String("ephemeral1"),
 				},
 			},
+
+			encrypted: false,
 		},
 	}
 
 	for i, tt := range amiTests {
 		t.Run(fmt.Sprintf("%d: %s", i, tt.description), func(t *testing.T) {
 			mockProvider := mockDescribeImages(tt.blockDeviceMappings, tt.rootDeviceName)
-			err := ami.Use(mockProvider.MockEC2(), &api.NodeGroup{
+			ng := &api.NodeGroup{
 				AMI: "ami-0121d8347f8191f90",
-			})
+			}
+			err := ami.Use(mockProvider.MockEC2(), ng)
 
 			if err != nil {
 				t.Errorf("unexpected error: %v", err)
+			}
+
+			if *ng.VolumeEncrypted != tt.encrypted {
+				t.Errorf("expected VolumeEncrypted to be %v", tt.encrypted)
 			}
 		})
 	}

--- a/pkg/ami/ami_test.go
+++ b/pkg/ami/ami_test.go
@@ -1,0 +1,77 @@
+package ami
+
+import (
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/stretchr/testify/mock"
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
+	"github.com/weaveworks/eksctl/pkg/testutils/mockprovider"
+)
+
+func TestUseAMI(t *testing.T) {
+	amiTests := []struct {
+		blockDeviceMappings []*ec2.BlockDeviceMapping
+		rootDeviceName      string
+	}{
+		{
+			rootDeviceName: "/dev/sda1",
+			blockDeviceMappings: []*ec2.BlockDeviceMapping{
+				{
+					DeviceName:  aws.String("xvdca"),
+					VirtualName: aws.String("ephemeral0"),
+				},
+				{
+					DeviceName:  aws.String("xvdcb"),
+					VirtualName: aws.String("ephemeral1"),
+				},
+				{
+					DeviceName: aws.String("/dev/sda1"),
+					Ebs: &ec2.EbsBlockDevice{
+						Encrypted: aws.Bool(true),
+					},
+				},
+			},
+		},
+	}
+
+	for i, tt := range amiTests {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			mockProvider := mockDescribeImages(tt.blockDeviceMappings, tt.rootDeviceName)
+			err := Use(mockProvider.MockEC2(), &api.NodeGroup{
+				AMI: "ami-0121d8347f8191f90",
+			})
+
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+		})
+	}
+
+}
+
+func mockDescribeImages(blockDeviceMappings []*ec2.BlockDeviceMapping, rootDeviceName string) *mockprovider.MockProvider {
+	mockProvider := mockprovider.NewMockProvider()
+
+	mockProvider.MockEC2().On("DescribeImages", mock.MatchedBy(func(input *ec2.DescribeImagesInput) bool {
+		return len(input.ImageIds) == 1 && strings.HasPrefix(*input.ImageIds[0], "ami-")
+	})).Return(func(input *ec2.DescribeImagesInput) *ec2.DescribeImagesOutput {
+		return &ec2.DescribeImagesOutput{
+			Images: []*ec2.Image{
+				{
+					ImageId:             input.ImageIds[0],
+					RootDeviceName:      aws.String(rootDeviceName),
+					RootDeviceType:      aws.String("ebs"),
+					BlockDeviceMappings: blockDeviceMappings,
+				},
+			},
+		}
+	}, func(*ec2.DescribeImagesInput) error {
+		return nil
+	})
+
+	return mockProvider
+}

--- a/pkg/ami/api.go
+++ b/pkg/ami/api.go
@@ -53,8 +53,12 @@ func Use(ec2api ec2iface.EC2API, ng *api.NodeGroup) error {
 		if !api.IsSetAndNonEmptyString(ng.VolumeName) {
 			ng.VolumeName = image.RootDeviceName
 		}
+		rootDeviceMapping, err := findRootDeviceMapping(image)
+		if err != nil {
+			return err
+		}
 
-		amiEncrypted := image.BlockDeviceMappings[0].Ebs.Encrypted
+		amiEncrypted := rootDeviceMapping.Ebs.Encrypted
 		if ng.VolumeEncrypted == nil {
 			ng.VolumeEncrypted = amiEncrypted
 		} else {
@@ -66,6 +70,15 @@ func Use(ec2api ec2iface.EC2API, ng *api.NodeGroup) error {
 	}
 
 	return nil
+}
+
+func findRootDeviceMapping(image *ec2.Image) (*ec2.BlockDeviceMapping, error) {
+	for _, deviceMapping := range image.BlockDeviceMappings {
+		if *deviceMapping.DeviceName == *image.RootDeviceName {
+			return deviceMapping, nil
+		}
+	}
+	return nil, errors.Errorf("failed to find root device mapping for AMI %q", *image.ImageId)
 }
 
 // FindImage will get the AMI to use for the EKS nodes by querying AWS EC2 API.

--- a/pkg/eks/api_test.go
+++ b/pkg/eks/api_test.go
@@ -204,8 +204,10 @@ func mockDescribeImages(p *mockprovider.MockProvider, amiId string, matcher func
 					State:          aws.String("available"),
 					OwnerId:        aws.String("123"),
 					RootDeviceType: aws.String("ebs"),
+					RootDeviceName: aws.String("/dev/sda1"),
 					BlockDeviceMappings: []*ec2.BlockDeviceMapping{
 						{
+							DeviceName: aws.String("/dev/sda1"),
 							Ebs: &ec2.EbsBlockDevice{
 								Encrypted: aws.Bool(false),
 							},

